cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll

```diff
@@ -796,10 +799,52 @@ private Node getAPriorDefinition(SsaDefOrUse defOrUse) {
   )
 }
 
+private predicate inOut(FIO::FunctionInput input, FIO::FunctionOutput output) {
+  exists(int indirectionIndex |
+    input.isQualifierObject(indirectionIndex) and
+    output.isQualifierObject(indirectionIndex)
+    or
+    exists(int i |
+      input.isParameterDeref(i, indirectionIndex) and
+      output.isParameterDeref(i, indirectionIndex)
+    )
+  )
+}
+
+/**
+ * Holds if there should not be use-use flow out of `n` (or a conversion that
+ * flows to `n`).
+ */
+private predicate modeledFlowBarrier(Node n) {
+  exists(
+    FIO::FunctionInput input, FIO::FunctionOutput output, CallInstruction call,
+    PartialFlow::PartialFlowFunction partialFlowFunc
+  |
+    n = callInput(call, input) and
+    inOut(input, output) and
+    exists(callOutput(call, output)) and
+    partialFlowFunc = call.getStaticCallTarget() and
+    not partialFlowFunc.isPartialWrite(output)
```

Comment:
Two points here:
- Generally, I like to keep negations simple for performance reasons. Since the body of a negation is always materialized, if the body of the negation is simply a call to a predicate then this is _already_ materialized. However, if we change it to `call.getStaticCallTarget().(PartialFlow::PartialFlowFunction).isPartialWrite(output)` then the compiler will have to materialize this predicate:
  ```ql
  modeledFlowBarrier_antijoin(CallInstruction call, FIO::FunctionOutput output) {
    call.getStaticCallTarget().(PartialFlow::PartialFlowFunction).isPartialWrite(output)
  }
  ```
  whereas the current implementation doesn't force the compiler to do this since the body of the negation is already just a call to `isPartialWrite`. So the antijoin would look like:
  ```ql
  modeledFlowBarrier_antijoin(PartialFlow::PartialFlowFunction partialFlowFunc, FIO::FunctionOutput output) {
    partialFlowFunc.isPartialWrite(output)
  }
  ```
  and the compiler simply optimizes out that predicate. This is why negation bodies should always be kept small when possible.
- Your suggestion actually subtly changes the semantics of `modeledFlowBarrier`! After your change, any call to a function that doesn't have a call target will satisfy `not call.getStaticCallTarget().(PartialFlow::PartialFlowFunction).isPartialWrite(output)`, and thus all arguments to function pointer calls will be seen as flow barriers. That's definitely not what we want.

What we can do is something like:
```ql
exists(..., Function func |
  ...
  func = call.getStaticCallTarget() and
  not func.(PartialFlow::PartialFlowFunction).isPartialWrite(output)
)
```
this ensures that we have the same behavior as what's currently in the PR, and that this won't silently change in the future if we change the fact that not all `DataFlowFunction`s and `TaintFunction`s are `PartialFlowFunction`s.

Before accepting that change, though, I'd like to verify that the compiler doesn't generate an unnecessarily materialized antijoin predicate.

Reply:
Yes, I had thought it would be a simple change that would improve robustness but it's clearly a tradeoff with performance and/or complexity.  Lets leave it as it is.